common/windows/win32_window.cc

```diff
@@ -0,0 +1,150 @@
+  return static_cast<int>(static_cast<double>(source * scale_factor));
+}
+
+Win32Window::Win32Window() {}
+
+Win32Window::~Win32Window() { Destroy(); }
+
+bool Win32Window::CreateAndShow(const char *title, const unsigned int x,
+                                const unsigned int y, const unsigned int width,
+                                const unsigned int height) {
+  Destroy();
+
+  WNDCLASS window_class = ResgisterWindowClass(title);
+
+  // New windows are created on the default monitor.  Window sizes are specified
+  // to the OS in physical pixels, hence to ensure a consistent size to will
+  // treat the width height passed in to this function as logical pixels and
+  // scale to appropriate for the default monitor.
+  HMONITOR defaut_mon = MonitorFromWindow(nullptr, MONITOR_DEFAULTTOPRIMARY);
+  UINT dpi_x = 0, dpi_y = 0;
+  GetDpiForMonitor(defaut_mon, MDT_EFFECTIVE_DPI, &dpi_x, &dpi_y);
+
+  double scaleFactor = static_cast<double>(dpi_x) / base_dpi;
+
+  auto window = CreateWindow(
+      window_class.lpszClassName, title, WS_OVERLAPPEDWINDOW | WS_VISIBLE,
+      Scale(x, scaleFactor), Scale(y, scaleFactor), Scale(width, scaleFactor),
+      Scale(height, scaleFactor), nullptr, nullptr, window_class.hInstance,
+      this);
+  if (window == nullptr) {
+    return false;
+  }
+  return true;
+}
+
+WNDCLASS Win32Window::ResgisterWindowClass(const char *title) {
+  window_class_name_ = title;
+
+  WNDCLASS window_class{};
+  window_class.hCursor = LoadCursor(nullptr, IDC_ARROW);
+  window_class.lpszClassName = "CLASSNAME";
+  window_class.style = CS_HREDRAW | CS_VREDRAW;
+  window_class.cbClsExtra = 0;
+  window_class.cbWndExtra = 0;
+  window_class.hInstance = GetModuleHandle(nullptr);
+  window_class.hIcon = nullptr;
+  window_class.hbrBackground = 0;
+  window_class.lpszMenuName = nullptr;
+  window_class.lpfnWndProc = WndProc;
+  RegisterClass(&window_class);
+  return window_class;
+}
+
+LRESULT CALLBACK Win32Window::WndProc(HWND const window, UINT const message,
+                                      WPARAM const wparam,
+                                      LPARAM const lparam) noexcept {
+  if (message == WM_NCCREATE) {
+    auto cs = reinterpret_cast<CREATESTRUCT *>(lparam);
+    SetWindowLongPtr(window, GWLP_USERDATA,
+                     reinterpret_cast<LONG_PTR>(cs->lpCreateParams));
+
+    auto that = static_cast<Win32Window *>(cs->lpCreateParams);
+
+    that->window_handle_ = window;
+  } else if (Win32Window *that = GetThisFromHandle(window)) {
+    return that->MessageHandler(window, message, wparam, lparam);
+  }
+
+  return DefWindowProc(window, message, wparam, lparam);
+}
+
+LRESULT
+Win32Window::MessageHandler(HWND hwnd, UINT const message, WPARAM const wparam,
+                            LPARAM const lparam) noexcept {
+  int xPos = 0, yPos = 0;
+  UINT width = 0, height = 0;
+  auto window =
+      reinterpret_cast<Win32Window *>(GetWindowLongPtr(hwnd, GWLP_USERDATA));
+
+  if (window != nullptr) {
```

Comment:
I'd prefer early return on `nullptr` rather than deeper nesting on the bulk of the logic.